src/timer.rs

```diff
@@ -288,3 +281,502 @@ where
+        TimerDriver {
+            callback: Callback {
+                now: &self.current_time,
+            },
+            context: &self,
+        }
+    }
+}
+
+/// Timer driver instance. You can create a TimerDriver from a DriverContext as follows:
+/// ```no_run
+/// # use libtock::timer::DriverContext;
+/// # use libtock::result::TockResult;
+/// # #[libtock::main]
+/// # async fn main() -> TockResult<()> {
+/// let context = DriverContext::create()?;
+/// context.create_timer_driver().expect("The timer driver is a singleton and can only created once.");
+/// # Ok(())
+/// # }
+/// ```
+pub struct TimerDriver<'a> {
+    callback: Callback<'a>,
+    context: &'a DriverContext,
+}
+
+struct Callback<'a> {
+    now: &'a Cell<usize>,
+}
+
+impl<'a> SubscribableCallback for Callback<'a> {
+    fn call_rust(&mut self, now: usize, _: usize, _: usize) {
+        self.now.set(now);
+    }
+}
+
+/// Activated time driver. Updates current time in the context and manages
+/// active alarms.
+/// Example usage (sleep for 1 second):
+/// ```no_run
+/// # use libtock::timer::DriverContext;
+/// # use libtock::result::TockResult;
+/// # use libtock::timer::Duration;
+/// # #[libtock::main]
+/// # async fn main() -> TockResult<()> {
+/// let context = DriverContext::create()?;
+/// let mut driver = context.create_timer_driver().unwrap();
+/// let timer_driver = driver.activate()?;
+/// timer_driver.sleep(Duration::from_ms(1000)).await?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct ParallelSleepDriver<'a> {
+    _callback_subscription: CallbackSubscription<'a>,
+    context: &'a DriverContext,
+}
+
+impl<'a> TimerDriver<'a> {
+    /// Activate the timer driver, will return a ParallelSleepDriver which
+    /// can used to sleep.
+    pub fn activate(&'a mut self) -> TockResult<ParallelSleepDriver<'a>> {
+        let subscription = syscalls::subscribe(
+            DRIVER_NUMBER,
+            subscribe_nr::SUBSCRIBE_CALLBACK,
+            &mut self.callback,
+        )?;
+        let driver = ParallelSleepDriver {
+            _callback_subscription: subscription,
+            context: &self.context,
+        };
+        Ok(driver)
+    }
+}
+
+impl<'a> ParallelSleepDriver<'a> {
+    /// Sleep for the given duration
+    pub async fn sleep(&self, duration: Duration<usize>) -> TockResult<()> {
+        let now = get_current_ticks()?;
+        self.context.current_time.set(now);
+        let freq = get_clock_frequency()?;
+        let alarm_instant = Self::compute_alarm_instant(duration.ms, now, freq)?;
+        let this_alarm = ActiveTimer {
+            instant: alarm_instant as u32,
+            set_at: now as u32,
+        };
+
+        let suspended_timer: Cell<Option<ActiveTimer>> = Cell::new(None);
+
+        futures::wait_until(|| {
+            let now = match get_current_ticks() {
+                Ok(value) => value,
+                Err(_) => return false,
+            };
```

Comment:
It looks like most of the matches in this lambda could use the `?` early-return pattern, for example if the lambda is refactored as a thin wrapper around a function call, with a function returning `TockResult<()>` (or `TockResult<bool>`).

Reply:
I was thinking about something like this. I just don't know how to handle the results in `wait_until`.

Reply:
What I meant is that the lambda within `wait_until` could wrap a function call and do a single match or `unwrap_or` instead. Something like:

```rust
futures::wait_until(|| {
    self.wait_for_timer(/* parameters */).unwrap_or(false)
}

fn wait_for_timer(&self, /* parameters */) -> TockResult<bool> {
    /* current implementation, but using "?" instead of the matches */
}
```

Reply:
Fixed.